webapp/src/components/modals/create_issue/index.tsx

```diff
@@ -0,0 +1,280 @@
+import React, {useCallback, useEffect, useMemo, useState} from 'react';
+import {useDispatch, useSelector} from 'react-redux';
+import {Modal} from 'react-bootstrap';
+import {Theme} from 'mattermost-redux/types/preferences';
+import {getPost} from 'mattermost-redux/selectors/entities/posts';
+
+import {getErrorMessage} from 'src/utils/user_utils';
+import GitlabLabelSelector from 'src/components/gitlab_label_selector';
+import GitlabAssigneeSelector from 'src/components/gitlab_assignee_selector';
+import GitlabMilestoneSelector from 'src/components/gitlab_milestone_selector';
+import GitlabProjectSelector from 'src/components/gitlab_project_selector';
+import Validator from 'src/components/validator';
+import FormButton from 'src/components/form_button';
+import Input from 'src/components/input';
+import {id as pluginId} from 'src/manifest';
+import {closeCreateIssueModal, createIssue} from 'src/actions';
+import {GlobalState} from 'src/types/global_state';
+import {usePrevious} from 'src/hooks/use_previous';
+
+const MAX_TITLE_LENGTH = 255;
+
+type PropTypes = {
+    theme: Theme;
+};
+
+const CreateIssueModal = ({theme}: PropTypes) => {
+    const validator = useMemo(() => (new Validator()), []);
+    const [submitting, setSubmitting] = useState<boolean>(false);
+    const [error, setError] = useState<string | null>(null);
+    const [project, setProject] = useState<ProjectSelection | null>(null);
+    const [issueTitle, setIssueTitle] = useState<string>('');
+    const [issueDescription, setIssueDescription] = useState<string>('');
+    const [labels, setLabels] = useState<SelectionType[]>([]);
+    const [assignees, setAssignees] = useState<SelectionType[]>([]);
+    const [milestone, setMilestone] = useState<SelectionType | null>(null);
+    const [showErrors, setShowErrors] = useState<boolean>(false);
+    const [issueTitleValid, setIssueTitleValid] = useState<boolean>(true);
+
+    const {visible, post, channelId, title} = useSelector((state: GlobalState) => {
+        const {postId, title, channelId} = state[`plugins-${pluginId}` as plugin].createIssueModal;
+        
+        const post = postId ? getPost(state, postId) : null;
+        return {
+            visible: state[`plugins-${pluginId}` as plugin].isCreateIssueModalVisible,
+            post,
+            title,
+            channelId,
+        };
+    });
+
+    const dispatch = useDispatch();
+    const prevPost = usePrevious(post);
+    const prevChannelId = usePrevious(channelId);
+    const prevTitle = usePrevious(title);
+
+    useEffect(() => {
+        if (post && !prevPost) {
+            setIssueDescription(post.message);
+        } else if (channelId && (channelId !== prevChannelId || title !== prevTitle)) {
+            setIssueTitle(title.substring(0, MAX_TITLE_LENGTH));
+        }
+    }, [channelId, title, post]);
+
+    // handle issue creation after form is populated
+    const handleCreate = useCallback(async (e: React.FormEvent<HTMLFormElement> | Event) => {
```

Comment:
Also I recommend not using hooks like `useCallback` for this sort of use case. This modal/component is usually not open/visible when a user is using Mattermost, so it doesn't need to be this optimized. IMO this just adds complexity and risk with all the dependency array stuff, whereas having regular functions does not have any down sides for the purpose of this PR.

The one time it's good to use these is when we are setting up props for a component that will render a lot of items (like a huge array), so it doesn't need to do its heavy lifting rendering all the time. The project selector is potentially like that, but the performance hit on that project select box would first and foremost be the number of DOM elements in the select box if we have a lot of elements. Which we can limit how many are shown ([example from Jira plugin](https://github.com/mattermost/mattermost-plugin-jira/blob/675961b88fa8a23bdb36ec4ee609bf664afb5ec9/webapp/src/components/react_select_setting.tsx#L107-L121)), if we receive complaints when people have over 1000 projects, but I think that's unnecessary for this feature. The only thing that I would say may be worth optimizing are the calls to `useMemo`, but even then, I don't think that's worth it due to the risk it brings.

If there are particular places where you think we need optimizations, please bring them up so we can discuss how to approach it. But I personally think that this feature does not require any of the optimizations.